tensorcircuit/shadows.py

```diff
@@ -0,0 +1,282 @@
+"""Classical Shadows base class with processing functions"""
+from typing import Optional, Sequence
+from string import ascii_letters as ABC
+from tensorcircuit.cons import backend
+from tensorcircuit.circuit import Circuit
+import numpy as np
+
+
+def shadow_snapshots(psi, pauli_strings, repeat: int = 1):
+    '''
+        ns: number of snapshots
+        nq: number of qubits
+        cir: quantum circuit
+        pauli_strings = (ns, nq)
+        repeat: times to measure on one pauli string
+
+        return:
+        snapshots = (ns, repeat, nq)
+    '''
+    if pauli_strings.dtype not in ("int", "int32"):
+        raise TypeError("Expected a int data type. " f"Got {pauli_strings.dtype}")
+
+    # if backend.max(pauli_strings) > 2 or backend.min(pauli_strings) < 0:
+    #     raise ValueError("The elements in pauli_strings must be between 0 and 2!")
+
+    angles = backend.cast(backend.convert_to_tensor(
+        np.array([[-np.pi / 2, np.pi / 4, 0], [np.pi / 3, np.arccos(1 / np.sqrt(3)), np.pi / 4], [0, 0, 0]])),
+                          dtype=pauli_strings.dtype)
+
+    nq = pauli_strings.shape[1]
+    assert 2 ** nq == len(psi)
+
+    def proj_measure(pauli_string):
+        # pauli_rot = backend.onehot(pauli_string, num=3) @ angles
+        pauli_rot = angles[pauli_string]
+        c_ = Circuit(nq, inputs=psi)
+        for i, (theta, alpha, phi) in enumerate(pauli_rot):
+            c_.R(i, theta=theta, alpha=alpha, phi=phi)
+        return c_.sample(batch=repeat, format="sample_bin")
```

Comment:
add argument: `allow_state=True` or allow this argument be passed in from outside function